File: android-server/app/src/main/java/com/ycdesk/server/SignalingServer.java
package com.ycdesk.server;

import android.util.Log;
import org.nanohttpd.protocols.http.IHTTPSession;
import org.nanohttpd.protocols.http.response.Response;
import org.nanohttpd.protocols.websockets.NanoWSD;
import org.nanohttpd.protocols.websockets.WebSocket;
import org.nanohttpd.protocols.websockets.WebSocketFrame;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import org.json.JSONObject;
import org.json.JSONException;

public class SignalingServer extends NanoWSD {

    private static final String TAG = "SignalingServer";
    
    private final int port;
    private final boolean useHttps;
    
    private final Map<String, WebSocket> deviceSockets = new ConcurrentHashMap<>();
    private final Map<String, String> socketToDevice = new ConcurrentHashMap<>();
    private final Map<String, SessionInfo> sessions = new ConcurrentHashMap<>();

    public SignalingServer(int port, boolean useHttps) {
        super(port);
        this.port = port;
        this.useHttps = useHttps;
        Log.d(TAG, "SignalingServer created on port: " + port);
    }

    @Override
    protected WebSocket openWebSocket(IHTTPSession handshake) {
        return new SignalingWebSocket(handshake);
    }

    private class SignalingWebSocket extends WebSocket {

        private String deviceId;

        public SignalingWebSocket(IHTTPSession handshakeRequest) {
            super(handshakeRequest);
        }

        @Override
        protected void onOpen() {
            Log.d(TAG, "WebSocket opened");
        }

        @Override
        protected void onClose(WebSocketFrame.CloseCode code, String reason, boolean initiatedByRemote) {
            Log.d(TAG, "WebSocket closed: " + reason);
            if (deviceId != null) {
                deviceSockets.remove(deviceId);
                socketToDevice.remove(this.hashCode());
            }
        }

        @Override
        protected void onMessage(WebSocketFrame message) {
            try {
                String payload = message.getTextPayload();
                Log.d(TAG, "Received: " + payload);
                JSONObject json = new JSONObject(payload);
                handleMessage(json);
            } catch (Exception e) {
                Log.e(TAG, "Error handling message", e);
            }
        }

        @Override
        protected void onPong(WebSocketFrame pong) {
        }

        @Override
        protected void onException(IOException exception) {
            Log.e(TAG, "WebSocket exception", exception);
        }

        private void handleMessage(JSONObject json) throws JSONException {
            String type = json.optString("type");
            
            switch (type) {
                case "register":
                    handleRegister(json);
                    break;
                case "connect-request":
                    handleConnectRequest(json);
                    break;
                case "connection-response":
                    handleConnectionResponse(json);
                    break;
                case "offer":
                    handleOffer(json);
                    break;
                case "answer":
                    handleAnswer(json);
                    break;
                case "ice-candidate":
                    handleIceCandidate(json);
                    break;
            }
        }

        private void handleRegister(JSONObject json) throws JSONException {
            deviceId = json.getString("deviceId");
            deviceSockets.put(deviceId, this);
            socketToDevice.put(this.hashCode(), deviceId);
            Log.d(TAG, "Device registered: " + deviceId);
        }

        private void handleConnectRequest(JSONObject json) throws JSONException {
            String fromDeviceId = json.getString("fromDeviceId");
            String toDeviceId = json.getString("toDeviceId");
            
            WebSocket targetSocket = deviceSockets.get(toDeviceId);
            
            if (targetSocket != null) {
                String sessionId = generateSessionId();
                sessions.put(sessionId, new SessionInfo(fromDeviceId, toDeviceId));
                
                JSONObject response = new JSONObject();
                response.put("type", "incoming-connection");
                response.put("fromDeviceId", fromDeviceId);
                response.put("sessionId", sessionId);
                sendMessage(targetSocket, response);
                
                Log.d(TAG, "Connect request: " + fromDeviceId + " -> " + toDeviceId);
            } else {
                JSONObject response = new JSONObject();
                response.put("type", "connection-result");
                response.put("accepted", false);
                response.put("error", "Device not online");
                sendMessage(this, response);
            }
        }

        private void handleConnectionResponse(JSONObject json) throws JSONException {
            String sessionId = json.getString("sessionId");
            boolean accepted = json.getBoolean("accepted");
            String fromDeviceId = json.getString("fromDeviceId");
            String toDeviceId = json.getString("toDeviceId");
            
            SessionInfo session = sessions.get(sessionId);
            if (session != null) {
                WebSocket targetSocket = deviceSockets.get(fromDeviceId);
                if (targetSocket != null) {
                    JSONObject response = new JSONObject();
                    response.put("type", "connection-result");
                    response.put("accepted", accepted);
                    response.put("sessionId", sessionId);
                    response.put("fromDeviceId", session.fromDeviceId);
                    response.put("toDeviceId", session.toDeviceId);
                    sendMessage(targetSocket, response);
                }
            }
        }

        private void handleOffer(JSONObject json) throws JSONException {
            String sessionId = json.getString("sessionId");
            JSONObject offer = json.getJSONObject("offer");
            String toDeviceId = json.getString("toDeviceId");
            
            WebSocket targetSocket = deviceSockets.get(toDeviceId);
            if (targetSocket != null) {
                JSONObject response = new JSONObject();
                response.put("type", "offer");
                response.put("sessionId", sessionId);
                response.put("offer", offer);
                sendMessage(targetSocket, response);
            }
        }

        private void handleAnswer(JSONObject json) throws JSONException {
            String sessionId = json.getString("sessionId");
            JSONObject answer = json.getJSONObject("answer");
            String toDeviceId = json.getString("toDeviceId");
            
            WebSocket targetSocket = deviceSockets.get(toDeviceId);
            if (targetSocket != null) {
                JSONObject response = new JSONObject();
                response.put("type", "answer");
                response.put("sessionId", sessionId);
                response.put("answer", answer);
                sendMessage(targetSocket, response);
            }
        }

        private void handleIceCandidate(JSONObject json) throws JSONException {
            String sessionId = json.getString("sessionId");
            JSONObject candidate = json.getJSONObject("candidate");
            String toDeviceId = json.getString("toDeviceId");
            
            WebSocket targetSocket = deviceSockets.get(toDeviceId);
            if (targetSocket != null) {
                JSONObject response = new JSONObject();
                response.put("type", "ice-candidate");
                response.put("sessionId", sessionId);
                response.put("candidate", candidate);
                sendMessage(targetSocket, response);
            }
        }

        private void sendMessage(WebSocket socket, JSONObject json) {
            try {
                socket.send(json.toString());
            } catch (IOException e) {
                Log.e(TAG, "Error sending message", e);
            }
        }
    }

    private String generateSessionId() {
        return Long.toHexString(System.currentTimeMillis()).toUpperCase();
    }

    private static class SessionInfo {
        String fromDeviceId;
        String toDeviceId;

        SessionInfo(String from, String to) {
            this.fromDeviceId = from;
            this.toDeviceId = to;
        }
    }
}

File: android-server/app/src/main/java/com/ycdesk/server/SignalingServer.kt
package com.ycdesk.server

import android.util.Log
import org.java_websocket.WebSocket
import org.java_websocket.handshake.ClientHandshake
import org.java_websocket.server.WebSocketServer
import org.json.JSONObject
import java.net.InetSocketAddress
import java.util.Date
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

class SignalingServer(private val port: Int, private val useHttps: Boolean) {

    companion object {
        private const val TAG = "SignalingServer"
        private const val SESSION_TIMEOUT_MS = 300000L
    }

    private var server: WebSocketServer? = null
    private val devices = ConcurrentHashMap<String, WebSocket>()
    private val sessions = ConcurrentHashMap<String, SessionInfo>()
    private val scheduler = Executors.newSingleThreadScheduledExecutor()

    var logCallback: ((String) -> Unit)? = null
    var deviceCountCallback: ((Int) -> Unit)? = null

    private fun log(message: String) {
        Log.d(TAG, message)
        logCallback?.invoke(message)
    }

    init {
        log("SignalingServer created on port: $port")
        startSessionCleanup()
    }

    private fun startSessionCleanup() {
        scheduler.scheduleAtFixedRate({
            try {
                val now = Date()
                sessions.entries.removeIf { (_, session) ->
                    val age = now.time - session.createdAt.time
                    if (age > SESSION_TIMEOUT_MS) {
                        log("清理过期会话")
                        true
                    } else {
                        false
                    }
                }
            } catch (e: Exception) {
                log("清理会话错误: ${e.message}")
            }
        }, 1, 1, TimeUnit.MINUTES)
    }

    fun start() {
        try {
            log("创建WebSocket服务器...")

            server = object : WebSocketServer(InetSocketAddress(port)) {
                override fun onOpen(conn: WebSocket?, handshake: ClientHandshake?) {
                    log("客户端连接: ${conn?.remoteSocketAddress}")
                }

                override fun onClose(conn: WebSocket?, code: Int, reason: String?, remote: Boolean) {
                    log("客户端断开: ${conn?.remoteSocketAddress}, 原因: $reason")

                    val deviceToRemove = devices.entries.find { it.value == conn }
                    if (deviceToRemove != null) {
                        devices.remove(deviceToRemove.key)
                        log("设备断开: ${deviceToRemove.key}")
                        deviceCountCallback?.invoke(devices.size)
                    }
                }

                override fun onMessage(conn: WebSocket?, message: String?) {
                    if (conn == null || message == null) return

                    log("收到消息: ${message.take(200)}")

                    try {
                        val json = JSONObject(message)
                        val type = json.optString("type")

                        when (type) {
                            "register" -> handleRegister(conn, json)
                            "connect-request" -> handleConnectRequest(conn, json)
                            "connection-response" -> handleConnectionResponse(conn, json)
                            "offer" -> handleOffer(conn, json)
                            "answer" -> handleAnswer(conn, json)
                            "ice-candidate" -> handleIceCandidate(conn, json)
                        }
                    } catch (e: Exception) {
                        log("解析消息失败: ${e.message}")
                        e.printStackTrace()
                    }
                }

                override fun onError(conn: WebSocket?, ex: Exception?) {
                    log("WebSocket错误: ${ex?.message}")
                    ex?.printStackTrace()
                }

                override fun onStart() {
                    log("服务器启动成功，端口: $port")
                }
            }

            log("启动服务器监听...")
            server?.start()
            log("服务器已启动")
        } catch (e: Exception) {
            log("启动服务器失败: ${e.message}")
            e.printStackTrace()
            throw e
        }
    }

    private fun handleRegister(conn: WebSocket, json: JSONObject) {
        val deviceId = json.optString("deviceId", "")
        if (deviceId.isNotEmpty()) {
            devices[deviceId] = conn
            log("设备注册: $deviceId")
            send(conn, JSONObject().apply {
                put("type", "registered")
                put("deviceId", deviceId)
            })
            deviceCountCallback?.invoke(devices.size)
        }
    }

    private fun handleConnectRequest(conn: WebSocket, json: JSONObject) {
        val fromDeviceId = json.optString("fromDeviceId", "")
        val toDeviceId = json.optString("toDeviceId", "")

        log("连接请求: $fromDeviceId -> $toDeviceId")

        val targetDevice = devices[toDeviceId]

        if (targetDevice != null) {
            val sessionId = generateSessionId()
            sessions[sessionId] = SessionInfo(fromDeviceId, toDeviceId, Date())

            send(targetDevice, JSONObject().apply {
                put("type", "incoming-connection")
                put("fromDeviceId", fromDeviceId)
                put("sessionId", sessionId)
            })

            log("发送incoming-connection到 $toDeviceId, 会话: $sessionId")
        } else {
            log("目标设备不在线: $toDeviceId")
            send(conn, JSONObject().apply {
                put("type", "connection-failed")
                put("reason", "device-offline")
                put("toDeviceId", toDeviceId)
            })
        }
    }

    private fun handleConnectionResponse(conn: WebSocket, json: JSONObject) {
        val sessionId = json.optString("sessionId", "")
        val accepted = json.optBoolean("accepted", false)

        log("连接响应: $sessionId -> ${if (accepted) "accepted" else "rejected"}")

        val session = sessions[sessionId]
        if (session != null) {
            session.status = if (accepted) "active" else "rejected"

            val fromDevice = devices[session.fromDeviceId]
            if (fromDevice != null) {
                send(fromDevice, JSONObject().apply {
                    put("type", "connection-result")
                    put("accepted", accepted)
                    put("sessionId", sessionId)
                    put("fromDeviceId", session.fromDeviceId)
                    put("toDeviceId", session.toDeviceId)
                })
            }
        }
    }

    private fun handleOffer(conn: WebSocket, json: JSONObject) {
        val sessionId = json.optString("sessionId", "")
        val offer = json.optJSONObject("offer")
        val toDeviceId = json.optString("toDeviceId", "")
        val fromDeviceId = findDeviceBySocket(conn)

        var resolvedSessionId = sessionId
        if (resolvedSessionId.isEmpty() && fromDeviceId != null) {
            for ((sid, session) in sessions) {
                if (session.fromDeviceId == fromDeviceId && session.status == "active") {
                    resolvedSessionId = sid
                    break
                }
            }
        }

        log("转发Offer: $resolvedSessionId $fromDeviceId -> $toDeviceId")

        if (toDeviceId.isNotEmpty()) {
            val targetDevice = devices[toDeviceId]
            if (targetDevice != null) {
                send(targetDevice, JSONObject().apply {
                    put("type", "offer")
                    put("sessionId", resolvedSessionId)
                    put("offer", offer)
                    put("fromDeviceId", fromDeviceId ?: "")
                })
            } else {
                log("转发Offer失败: 目标设备不在线 $toDeviceId")
            }
        }
    }

    private fun handleAnswer(conn: WebSocket, json: JSONObject) {
        val sessionId = json.optString("sessionId", "")
        val answer = json.optJSONObject("answer")
        val toDeviceId = json.optString("toDeviceId", "")
        val fromDeviceId = findDeviceBySocket(conn)

        log("转发Answer: $sessionId $fromDeviceId -> $toDeviceId")

        if (toDeviceId.isNotEmpty()) {
            val targetDevice = devices[toDeviceId]
            if (targetDevice != null) {
                send(targetDevice, JSONObject().apply {
                    put("type", "answer")
                    put("sessionId", sessionId)
                    put("answer", answer)
                    put("fromDeviceId", fromDeviceId ?: "")
                })
            }
        }
    }

    private fun handleIceCandidate(conn: WebSocket, json: JSONObject) {
        val sessionId = json.optString("sessionId", "")
        val candidate = json.optJSONObject("candidate")
        val toDeviceId = json.optString("toDeviceId", "")
        val fromDeviceId = findDeviceBySocket(conn)

        if (toDeviceId.isNotEmpty()) {
            val targetDevice = devices[toDeviceId]
            if (targetDevice != null) {
                send(targetDevice, JSONObject().apply {
                    put("type", "ice-candidate")
                    put("sessionId", sessionId)
                    put("candidate", candidate)
                    put("fromDeviceId", fromDeviceId ?: "")
                })
            }
        }
    }

    private fun send(conn: WebSocket, json: JSONObject) {
        if (conn.isOpen) {
            conn.send(json.toString())
        }
    }

    private fun findDeviceBySocket(conn: WebSocket): String? {
        for ((deviceId, socket) in devices) {
            if (socket == conn) {
                return deviceId
            }
        }
        return null
    }

    private fun generateSessionId(): String {
        val chars = "abcdefghijklmnopqrstuvwxyz0123456789"
        return (1..9)
            .map { chars[(Math.random() * chars.length).toInt()] }
            .joinToString("")
            .uppercase()
    }

    fun stop() {
        log("停止服务器...")
        scheduler.shutdown()
        try {
            server?.stop()
        } catch (e: Exception) {
            log("停止服务器错误: ${e.message}")
        }
        devices.clear()
        sessions.clear()
        log("服务器已停止")
    }

    fun getOnlineDeviceCount(): Int = devices.size

    private data class SessionInfo(
        val fromDeviceId: String,
        val toDeviceId: String,
        val createdAt: Date,
        var status: String = "pending"
    )
}
